Broadcast-Ext/OpenTok Live/SampleHandler.m

```diff
@@ -0,0 +1,215 @@
+    
+    _num_frames = 0;
+    [self destroyPixelBuffers];
+    [_broadcastHelper connect];
+}
+
+- (void)broadcastPaused {
+    // User has requested to pause the broadcast. Samples will stop being delivered.
+}
+
+- (void)broadcastResumed {
+    // User has requested to resume the broadcast. Samples delivery will resume.
+}
+
+- (void)broadcastFinished {
+    // User has requested to finish the broadcast.
+    [_broadcastHelper disconnect];
+}
+
+- (void) processPixelBuffer:(CVPixelBufferRef)pixelBuffer timeStamp:(CMTime)ts
+{
+    CIImage *ciImage = [CIImage imageWithCVPixelBuffer:pixelBuffer
+                                               options:nil];
+    //CGFloat imageWidth = CVPixelBufferGetWidth(pixelBuffer);
+    //CGFloat imageHeight = CVPixelBufferGetHeight(pixelBuffer);
+    CGFloat aspectRatio = 1;//imageHeight / imageWidth;
+    
+    ciImage = [self scaleFilterImage:ciImage
+                     withAspectRatio:aspectRatio scale:0.60];
+    
+    if(_pixelBufferPool == nil ||
+       CVPixelBufferGetWidth(pixelBuffer) != CVPixelBufferGetWidth(_pixelBuffer) ||
+       CVPixelBufferGetHeight(pixelBuffer) != CVPixelBufferGetHeight(_pixelBuffer))
+    {
+        [self destroyPixelBuffers];
+        [self createPixelBufferPoolWithWidth:ciImage.extent.size.width
+                                      height:ciImage.extent.size.height];
+        CVPixelBufferPoolCreatePixelBuffer(NULL, _pixelBufferPool, &_pixelBuffer);
+    }
+    
+    [_ciContext render:ciImage toCVPixelBuffer:_pixelBuffer];
+    
+    [self.videoCaptureConsumer consumeImageBuffer:_pixelBuffer
+                                      orientation:OTVideoOrientationUp
+                                        timestamp:ts
+                                         metadata:nil];
+}
+
+- (void)processSampleBuffer:(CMSampleBufferRef)sampleBuffer withType:(RPSampleBufferType)sampleBufferType {
+    
+    switch (sampleBufferType) {
+        case RPSampleBufferTypeVideo:
+        {
+            // Handle video sample buffer
+            
+            // Codec limits to avoid reaching over 50MB memory
+            // Tested on iPhone 8
+            // VP8  : Height = 800, Width = 450, fps = 10
+            // H264 : Height = 1068, Width = 600, fps = 15 or more
+            _num_frames++;
+            if (_num_frames % 3 != 0)
+                return;
+            
+            //            CFRetain(sampleBuffer);
+            //dispatch_async(_capture_queue, ^{
+            if([self->_broadcastHelper isConnected] && self->_capturing)
+            {
+                [self processPixelBuffer:CMSampleBufferGetImageBuffer(sampleBuffer)
+                               timeStamp:CMSampleBufferGetPresentationTimeStamp(sampleBuffer)];
+                //CFRelease(sampleBuffer);
+            }
+            //});
+        }
+            break;
+        case RPSampleBufferTypeAudioApp:
+        {
+            // Handle audio sample buffer for app audio
```

Comment:
Can't we process the audio sample here the same way as the microphone?
If it is easy, we can broadcast also apps with audio like games

Reply:
The InApp audio samples are different than regular mic samples( they have different length as well sample rate) and need changes to the audio driver. We can add this functionality in the near future.

Reply:
👍